src/decorate.ts

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 Google LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     https://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+'use strict';
+import path = require('path');
+import { JSONPath, visit } from 'jsonc-parser';
+import * as vscode from 'vscode';
+import XRegExp = require('xregexp');
+import { CombinedMessage, ComplexMessage, Literal, Message, MessageList, Metadata, Parser, Placeholder } from './messageParser';
+
+export const placeholderDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#ff6f00'
+	},
+	dark: {
+		color: '#fff9c4'
+	}
+});
+export const selectDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#6a1b9a'
+	},
+	dark: {
+		color: '#ce93d8'
+	}
+});
+export const pluralDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#0277bd'
+	},
+	dark: {
+		color: '#b3e5fc'
+	}
+});
+
+const selectRegex = /^([^\{\}]+\s*,\s*(?:select|gender)\s*,\s*(?:[^\{\}]*\{.*\})*)$/;
+const pluralRegex = /^[^\{\}]+\s*,\s*plural\s*,\s*(?:offset:\d+)?\s*(?:[^\{\} ]*?\s*\{.*\})$/;
+const placeholderNameRegex = /^[a-zA-Z][a-zA-Z_$0-9]*$/; //Must be able to translate to a (non-private) Dart variable
+const keyNameRegex = /^[a-zA-Z][a-zA-Z_0-9]*$/; //Must be able to translate to a (non-private) Dart method
+
+export class Decorator {
+	diagnostics = vscode.languages.createDiagnosticCollection("arb");
+
+	constructor(context?: vscode.ExtensionContext) {
+		context?.subscriptions.push(this.diagnostics);
+	}
+
+	decorate(editor: vscode.TextEditor, messageList: MessageList, errors: Literal[]): { diagnostics: vscode.Diagnostic[]; decorations: Map<vscode.TextEditorDecorationType, vscode.Range[]>; } | null {
+		// Prefill decorations map to avoid having old decoration hanging around
+		let decorationsMap = new Map<vscode.TextEditorDecorationType, vscode.Range[]>([
+			[placeholderDecoration, []],
+			[selectDecoration, []],
+			[pluralDecoration, []],
+		]);
+		let diagnosticsList: vscode.Diagnostic[] = [];
+
+		for (const error of errors) {
+			showErrorAt(error.start, error.end, error.value, vscode.DiagnosticSeverity.Error);
+		}
```

Comment:
Could errors and decorations (colouring) be split out? If you're parsing once and can pass the results into multiple classes, you could have a class doing only colouring and another doing only diagnostics which might keep things a little simpler.

Reply:
It would be nicest probably to have decoration and diagnostics as a visitor pattern - but I will save that for another PR... ;)